modules/core/test/image/testImageTemplateMatching.cpp

```diff
@@ -0,0 +1,392 @@
+    -h\n\
+       Print the help.\n\n", ipath.c_str());
+
+    if (badparam)
+      fprintf(stdout, "\nERROR: Bad parameter [%s]\n", badparam);
+  }
+
+  bool getOptions(int argc, const char **argv, std::string &ipath, bool &click,
+                  bool &doTemplateMatching)
+  {
+    const char *optarg_;
+    int c;
+    while ((c = vpParseArgv::parse(argc, argv, GETOPTARGS, &optarg_)) > 1) {
+
+      switch (c) {
+      case 'i':
+        ipath = optarg_;
+        break;
+      case 'h':
+        usage(argv[0], NULL, ipath);
+        return false;
+        break;
+      case 't':
+        doTemplateMatching = true;
+        break;
+
+      case 'c':
+          click = true;
+          break;
+      case 'd':
+        break;
+
+      default:
+        usage(argv[0], optarg_, ipath);
+        return false;
+        break;
+      }
+    }
+
+    if ((c == 1) || (c == -1)) {
+      // standalone param or error
+      usage(argv[0], NULL, ipath);
+      std::cerr << "ERROR: " << std::endl;
+      std::cerr << "  Bad argument " << optarg_ << std::endl << std::endl;
+      return false;
+    }
+
+    return true;
+  }
+
+  //Reference implementation for normalizedCorrelation() function
+  double normalizedCorrelation_gold(const vpImage<double> &I1, const vpImage<double> &I2)
+  {
+    if ((I1.getHeight() != I2.getHeight()) || (I1.getWidth() != I2.getWidth())) {
+        std::cerr << "Error: in vpImageTools::normalizedCorrelation(): "
+                  << "image dimensions mismatch." << std::endl;
+        throw vpException(vpException::dimensionError, "Dimension mismatch!");
+      }
+
+      double a = I1.getMeanValue();
+      double b = I2.getMeanValue();
+
+      double ab = 0.0;
+      double a2 = 0.0;
+      double b2 = 0.0;
+
+      for (unsigned int i = 0; i < I1.getHeight(); ++i)
+        for (unsigned int j = 0; j < I1.getWidth(); ++j) {
+          ab += (I1(i, j) - a) * (I2(i, j) - b);
+          a2 += vpMath::sqr(I1(i, j) - a);
+          b2 += vpMath::sqr(I2(i, j) - b);
+        }
+
+      return ab / sqrt(a2 * b2);
+  }
+
+  void templateMatching_gold(const vpImage<unsigned char> &I, const vpImage<unsigned char> &I_tpl,
+                             const unsigned int step_u, const unsigned int step_v,
+                             vpImage<double> &I_score)
+  {
+    unsigned int height_tpl = I_tpl.getHeight(), width_tpl = I_tpl.getWidth();
+    vpImage<double> I_double, I_tpl_double, I_cur;
+    vpImageConvert::convert(I, I_double);
+    vpImageConvert::convert(I_tpl, I_tpl_double);
+    I_score.resize(I.getHeight() - height_tpl, I.getWidth() - width_tpl, 0.0);
+
+    for (unsigned int i = 0; i < I.getHeight()-height_tpl; i += step_v) {
+      for (unsigned int j = 0; j < I.getWidth()-width_tpl; j += step_u) {
+        vpRect roi( vpImagePoint(i, j), vpImagePoint(i+height_tpl-1, j+width_tpl-1) );
+        vpImageTools::crop(I_double, roi, I_cur);
+
+        I_score[i][j] = normalizedCorrelation_gold(I_cur, I_tpl_double);
```

Comment:
Instead to implement `normalizedCorrelation_gold()` twice in this test but also in the previous one, why not adding an addition parameter to `normalizedCorrelation(..., bool gold=false)` that allows to use the non optimized version.
And inside this function ensure that if `gold = false`,  `VISP_HAVE_SSE2` is defined, otherwise use the gold version